sklearn/decomposition/nmf.py

```diff
@@ -55,55 +58,117 @@ def check_non_negative(X, whom):
+    n_components : integer
         The number of components desired in the approximation.
 
-    variant : None | 'a' | 'ar'
-        The variant of the NNDSVD algorithm.
-        Accepts None, 'a', 'ar'
-        None: leaves the zero entries as zero
-        'a': Fills the zero entries with the average of X
-        'ar': Fills the zero entries with standard normal random variates.
-        Default: None
+    init :  None | 'random' | 'nndsvd' | 'nndsvda' | 'nndsvdar'
+        Method used to initialize the procedure.
+        Default: 'nndsvdar' if n_components < n_features, otherwise 'random'.
+        Valid options:
+            'random': non-negative random matrices, scaled with:
+                sqrt(X.mean() / n_components)
+            'nndsvd': Nonnegative Double Singular Value Decomposition (NNDSVD)
+                initialization (better for sparseness)
+            'nndsvda': NNDSVD with zeros filled with the average of X
+                (better when sparsity is not desired)
+            'nndsvdar': NNDSVD with zeros filled with small random values
+                (generally faster, less accurate alternative to NNDSVDa
+                for when sparsity is not desired)
 
     eps: float
         Truncate all values less then this in output to zero.
 
-    random_state : numpy.RandomState | int, optional
-        The generator used to fill in the zeros, when using variant='ar'
-        Default: numpy.random
+    random_state : int seed, RandomState instance, or None (default)
+        Random number generator seed control, used in 'nndsvdar' and
+        'random' modes.
 
     Returns
     -------
 
-    (W, H) :
-        Initial guesses for solving X ~= WH such that
-        the number of columns in W is n_components.
+    W : array-like, shape (n_samples, n_components)
+        Initial guesses for solving X ~= W.H
+
+    H : array-like, shape (n_components, n_features)
+        Initial guesses for solving X ~= W.H
 
     References
     ----------
     C. Boutsidis, E. Gallopoulos: SVD based initialization: A head start for
     nonnegative matrix factorization - Pattern Recognition, 2008
-
     http://tinyurl.com/nndsvd
     """
     check_non_negative(X, "NMF initialization")
-    if variant not in (None, 'a', 'ar'):
-        raise ValueError("Invalid variant name")
+    n_samples, n_features = X.shape
 
+    if init is None:
+        if n_components < n_features:
+            init = 'nndsvd'
+        else:
+            init = 'random'
+
+    # Random initialization
+    if init == 'random':
+        avg = np.sqrt(X.mean() / n_components)
+        rng = check_random_state(random_state)
+        W = avg * rng.randn(n_samples, n_components)
+        # we do not write np.abs(W, out=W) to stay compatible with
+        # numpy 1.5 and earlier where the 'out' keyword is not
+        # supported as a kwarg on ufuncs
+        np.abs(W, W)
+        H = avg * rng.randn(n_components, n_features)
+        np.abs(H, H)
```

Comment:
NMF is biconvex so you can just initialize W by zero if you start by updating H. Or vice-versa. To be confirmed but I think this converges faster.